### ISSUE TITLE
Casing of enums

### ISSUE THREAD

**Issue by annevk** [tc39-transfer/proposal-intl-displaynames#32]
Is there a reason the type field does not follow the recommendations of https://w3ctag.github.io/design-principles/#casing-rules? Does JavaScript have different rules?

**Comment by sffc** [comment]
Agreed; also see #29

**Comment by FrankYFTang** [comment]
@annevk  could you please clarify/list which specific type field and which specific row in https://w3ctag.github.io/design-principles/#casing-rules are you referring to?

**Comment by FrankYFTang** [comment]
are you referring to only type: "date-symbol" instead of "dateSymbol", "date-field" instead of "dateField"? OR 

Do you ALSO think we need to change to use  "week-of-year" instead of "weekOfYear" and "day-period" instead of "dayPeriod"?

**Comment by sffc** [comment]
Keys should be camel case; string enum values should be kebab case. Except, for formatToParts, the strings are camel case for historical reasons.

**Comment by FrankYFTang** [comment]
but the string of the formatToParts could be the same as the enum value of another API. For example, "dayPeriod" and "timeZoneName" as type in Intl.DateTimeFormat.prototype.formatToParts could be the "code" of Intl.DisplayNames.of(). If we use 'day-period", "time-zone-name" in Intl.DisplayNames.of(), then we cannot directly pass that to Intl.DisplayNames.of()  in some case. So I suggest we keep using "dayPeriod", "weekOfYear" and "timeZoneName" in camel case and only change "date-field" and "date-symbol" to kebab case.

**Comment by ljharb** [comment]
That seems like it will end up with a more confusing world than just making everything camelCase. Is there a reason that path wasn’t/isn’t being taken?

**Comment by sffc** [comment]
Thanks for raising this issue about inconsistency between Intl.DateFormat.prototype.formatToParts and Intl.DisplayNames.  It's unfortunate that the older formatToParts strings are camel case and introduce this inconsistency.  On the one hand, I want to follow the style guide and use kebab case for all new strings.  On the other hand, I do think it would be nice if we had the invariant where you could pass a formatToParts type as the code for Intl.DisplayNames.  I can be convinced either way.

> That seems like it will end up with a more confusing world than just making everything camelCase. Is there a reason that path wasn’t/isn’t being taken?

Because of style guides such as the one @annevk posted in the OP.  @littledan has also been a big advocate for keeping strings as kebab case and keys as camel case.

**Comment by sffc** [comment]
@ljharb does bring up a good point.  Since we already have precedent for camel case in both properties and in strings, should we consider keeping that style, even if it is in conflict with some style guides?  I do not think we actually have kebab case in very many places yet.  @leobalter, do you have a sense for how common kebab case occurs in ECMA 402?  I know there is at least one instance of it in Unified NumberFormat (signDisplay: "except-zero"), but we could make a quick patch to that before it ships in any browser (we would need to act fast).

**Comment by ljharb** [comment]
kebab-case doesn't seem to appear anywhere in 262, so it seems like an odd choice to me for 402.

**Comment by leobalter** [comment]
@ljharb do we have any enumeration string values specified in 262?

The case here is for the values we use within options for 402's api. I tend to agree with the recommendation proposed by @annevk.

The kebab casing style is already present in one of our core parts for 402: `"en-GB-u-ca-islamic"`, "de-DE-u-co-phonebk", where it's _gracefully case sensitive_ and meaningful as values.

What really antagonizes with the kebab casing for string values is not camelCase but plain text as in `"best fit"` (case sensitive) for matchers. Other forms identified are `"America/New_York"` for timezone, which doesn't really match any of the styles mentioned so far, right?

Some kebab case is also found on the DateTimeFormat api: `"2-digit"`, but we also find lowercase concatenated values for `nu` (same for NumberFormat), e.g.: `"fullwide"` and `"arabext"`.

camelCase is still used for property names and that's not what we are discussing here, right?

I don't immediately remember of other values from 402. 

---

My conclusion: I'm highly in favor to follow the recommendation to use kebab-case for enumeration string values. We should document this somewhere.

As a side effect, we might discuss (as a separate issue in ecma402) also accepting a `"best-fit"` value for matchers to follow consistency. Of course not dropping compatibility for the already implemented `"best fit"`.

**Comment by ljharb** [comment]
Timezone names come from IANA, so that doesn’t seem like it need have any weight with respect to anything else. Locale names, i think, also come from cldr and as such don’t need to have weight either?

There are no observable strings I’m aware of in 262 besides property names, which are all in camelCase.

Surely it’s good for 402 to be consistent in the strings it gets to create, even when there’s “legacy” inconsistencies that need support - 262 is no stranger to that, see function lengths - and it’s yalls place to decide your target consistency. That said, it seems to me like separate from locale and timezone names, using camelCase would fit javascript idiom more closely.

**Comment by FrankYFTang** [comment]
Please also comment on https://github.com/tc39/proposal-intl-displaynames/pull/33

**Comment by leobalter** [comment]
> using camelCase would fit javascript idiom more closely.

I don't see how strings used only as values - and not as property names - will fit closer as camelCase. I just showed the examples of current values described in ECMA-402.

> Locale names, i think, also come from cldr and as such don’t need to have weight either?

I also just said I see them as core for the Intl specs, so we are disagreeing with the opposite directions here.

We need to find consensus anyway. I don't plan to oppose to camelCase if that ends up desired by everyone but I continue to highly recommend kebab-case for enum strings. 

1. If a string is described as a property key/name, camelCase must be used.
2.If a string is described as a {property,name} value, kebab-case must be used.

**Comment by leobalter** [comment]
I left a comment in another issue for something in the current proposed specs that would be affected by the recommendation we consent here. https://github.com/tc39/proposal-intl-displaynames/pull/33#issuecomment-503655672

**Comment by sffc** [comment]
I think the main open issue with regard to adopting kebab case for enum strings is the one Frank mentioned above:

> If we use 'day-period", "time-zone-name" in Intl.DisplayNames.of(), then we cannot directly pass that to Intl.DisplayNames.of() in some case.

One way around that would be to support *both* kebab case and camel case just for those two specific cases for better backwards compatibility.

**Comment by leobalter** [comment]
I'm fine with "aliasing" a name for backwards compat, so we have a repeated `"day-period"` and `"dayPeriod"`. One name means for consistency and initial recommendation, the other works like an annex B sort.

**Comment by littledan** [comment]
Aliasing seems like much more trouble than it's worth. I would prefer to stick with camelCase than add aliases. To summarize, the issue is that some type values are also used as property keys, and other type values are used as Intl.DisplayName enums, and so if we made the later kebab-case, it would be a jarring irregularity? I guess this is an OK reason to stick with camelCase, if @annevk is OK with it.

**Comment by annevk** [comment]
I think it might turn out to be equally jarring if builtin modules end up with different conventions down the line, more or less depending on who were involved in building them.

I could understand the precedence argument if JavaScript had established precedent.

**Comment by littledan** [comment]
@annevk Unfortunately, there's already precedent in ECMA-402 for camel-case for strings representing types of parts (this code works :
```
new Intl.NumberFormat('en').formatToParts(-3)[0].type // "minusSign" in V8 and SpiderMonkey
```
That one in isolation feels like a bit of gratuitous, and may have something to do with the people involved not being aware of the broader convention. But it gets more complicated with `dayPeriod`: In Intl, there are a number of strings used both as a property key of an object and as a type in an enum like this. For example, `dayPeriod` is one of them: It may be in four situations:
1. In the options bag passed into the `Intl.DateTimeFormat` constructor
1. In the `resolvedOptions` object
1. As the `type: ` value of one of the Array elements returned by `formatToParts`
1. Passed into an `Intl.DisplayNames` instance

I think it'd be pretty confusing to use `dayPeriod` for 1, 2 and 3, and `day-period` for 4, especially when 3 already "should" use `day-period`. Even if we say that we caught it early enough to iterate on the casing for 3, the wide shipping of `minusSign` would mean that we have an inconsistency within the `formatToParts` API with respect to the conventions. It will be pretty confusing, I think, to remember within this API when to use the string `"dayPeriod"` and when to use `"day-period"`, if we really split it out.

Given the situation, for consistency within the Intl API, I'd say we should stick with `camelCase` for all of these strings.

Unfortunately, here, we have a tradeoff between local consistency and global consistency. This tradeoff also exists when we violate the Web-wide convention of not putting a class/interface on a namespace object--we're continuing to put new constructors on the Intl object, because it'd be weird to have `Intl.DateTimeFormat` on one hand and `IntlDisplayNames` on the other.

But I'm encouraged by this thread, where we're able to have a nice conversation about Web/JS-wide conventions, with everyone taking this regularity seriously. This makes me optimistic for future APIs following conventions when possible.

**Comment by leobalter** [comment]
At the ECMA402 meeting today it was decided that camelCase prevails as the most consistent (or the least inconsistent) option. The consensus is to recommend use of camelCase even for enumeration string values. This also includes not using aliases in kebab-case and it won't change anything already shipping in the specs.

The recommendation becomes effective to this and any other active proposals (< stage 4) and any new proposal in the future. e.g. This will affect the Unified NumberFormat proposal from @sffc. 

In a personal note, the sentiment shared was that it would be easier - and perhaps better - to use kebab-case if we could do it from a blank canvas, but that's not the case.

I believe this issue is solved for now and I appreciate the insightful conversations. This affects the general aspects of the specs style and we hope to achieve a better consistency along the way.

**Comment by sffc** [comment]
@leobalter Is there a style guide somewhere we should document this decision, and the scope of the decision (for example, this affects options and string enums, but not syntaxes like BCP47 and unit identifiers)?

**Comment by sffc** [comment]
Reopening this issue based on discussion with TC39.  We failed to reach consensus one way or another.  I would like to create an explainer for this decision, summarizing the pros and cons, and then verify the conclusion.  I'll post that here and then I'd like review from the relevant stakeholders: @waldemarhorwat @wycats @domenic @gesa

**Comment by littledan** [comment]
You can find the notes from that discussion at https://github.com/tc39/tc39-notes/blob/master/meetings/2019-07/july-24.md#casing-conventions

**Comment by littledan** [comment]
I followed up with @wycats offline. It seems like there was no strong disagreement here, just a wish that the reasoning here be clearly explained to the committee.

I'd like to propose that we adopt camelCase as a convention, but explicitly scope this convention to ECMA-402. I didn't see this scoping made explicit in the presentation, and the ensuing discussion also seemed to contain this confusion. For compatibility reasons, we won't change existing strings like `"bestFit"` to kebab-case, or `"timed-out"` (from `Atomics`) to camelCase, so I think that a scoped convention is the best we can do given the situation we have.

Let's discuss whether this would be acceptable here on this thread, so we can figure out whether we could go back to TC39 in October 2019 with this proposed conclusion.

**Comment by leobalter** [comment]
it was scoped to ECMA-402 when we presented at TC39, and it's an error if people perceived it differently.

People at TC39 wanted some more participation in the reasoning, although no one engaged in the current threads after the meeting.

I'm not sure how to follow up on this. My only goal is to support unblocking the active proposals for Intl.

**Comment by littledan** [comment]
@leobalter That's unfortunate. Maybe we could try simply bringing it back to give just a little more time for discussion?

**Comment by sffc** [comment]
I've been procrastinating on making the explainer but it has been my intention to circulate it in advance of a follow up discussion at TC39 in October. I'll try to get to that today or tomorrow.

**Comment by sffc** [comment]
Here is the explainer:

https://github.com/tc39/ecma402/pull/377

**Comment by gesa** [comment]
The explainer is fantastic @sffc, crystal clear and easy to follow, thank you
